core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -648,11 +649,11 @@ private boolean allocateUnassigned() {
                 for (int i = 0; i < primaryLength; i++) {
                     ShardRouting shard = primary[i];
                     if (!shard.primary()) {
-                        boolean drop = deciders.canAllocate(shard, allocation).type() == Type.NO;
-                        if (drop) {
-                            unassigned.ignoreShard(shard);
+                        final Decision decision = deciders.canAllocate(shard, allocation);
+                        if (decision.type() == Type.NO) {
+                            changed |= unassigned.ignoreShard(shard, UnsuccessfulAllocationAttempt.DENIED);
```

Comment:
can you try adding a static helper method to unassignedinfo that converts Decision or Decision.Type to AllocationAttempt? this might make things simpler (we don't have to make sure to pass the right value)